blockchain/utxocache.go

```diff
@@ -0,0 +1,790 @@
+// in a block at a height less than the eviction height will be evicted from the
+// cache when the cache reaches its maximum allowed size.
+//
+// Eviction is based on height since the height of the block that an entry is
+// contained in is a proxy for how old the utxo is.  On average, recent utxos
+// are much more likely to be spent in upcoming blocks than older utxos, so the
+// strategy used is to evict the oldest utxos in order to maximize the hit ratio
+// of the cache.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) calcEvictionHeight(bestHeight uint32) uint32 {
+	if bestHeight < c.lastEvictionHeight {
+		return bestHeight
+	}
+
+	lastEvictionDepth := bestHeight - c.lastEvictionHeight
+	numBlocksToEvict := math.Ceil(float64(lastEvictionDepth) * evictionPercentage)
+	return c.lastEvictionHeight + uint32(numBlocksToEvict)
+}
+
+// shouldFlush returns whether or not a flush should be performed.
+//
+// If the maximum size of the cache has been reached, or if the periodic flush
+// duration has been reached, then a flush is required.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) shouldFlush(bestHash *chainhash.Hash) bool {
+	// No need to flush if the cache has already been flushed through the best
+	// hash.
+	if c.lastFlushHash == *bestHash {
+		return false
+	}
+
+	// Flush if the max size of the cache has been reached.
+	if c.totalSize() >= c.maxSize {
+		return true
+	}
+
+	// Flush if the periodic flush duration has been reached.
+	now := time.Now()
+	duration := now.Sub(c.lastFlushTime)
+	return duration >= time.Minute*periodicFlushMins
+}
+
+// flush commits all modified entries to the database and conditionally evicts
+// entries.
+//
+// Entries that are nil or spent are always evicted since they are
+// unlikely to be accessed again.  Additionally, if the cache has reached its
+// maximum size, entries are evicted based on the height of the block that they
+// are contained in.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) flush(bestHash *chainhash.Hash, bestHeight uint32, logFlush bool) error {
+	// If the maximum allowed size of the cache has been reached, determine the
+	// eviction height.
+	var evictionHeight uint32
+	memUsageMiB := float64(c.totalSize()) / 1024 / 1024
+	maxMemUsageMiB := float64(c.maxSize) / 1024 / 1024
+	if memUsageMiB >= maxMemUsageMiB {
+		evictionHeight = c.calcEvictionHeight(bestHeight)
+	}
+
+	// Log that a flush is starting and indicate the current memory usage, hit
+	// ratio, and eviction height.
+	var hitRatio float64
+	var evictionLog string
+	if logFlush {
+		memUsagePercent := memUsageMiB / maxMemUsageMiB * 100
+		hitRatio = c.hitRatio()
+		if evictionHeight != 0 {
+			evictionLog = fmt.Sprintf(", eviction height: %d", evictionHeight)
+		}
+		log.Infof("UTXO cache flush starting (%d entries, %.2f MiB (%.2f%%), "+
```

Comment:
Since `utxocachemaxsize` is the only relevant user-configurable variable in the sampleconfig. I think it would absolutely be valuable to be able to know if/when to increase the size.

Folks with extremely limited memory could also lower the `evictionPercentage` default to keep their hit rate high, but since the default only evicts 15% of the oldest blocks, it's hard to lower that much more.

Reply:
I updated this to log at the debug level.  The idea of indicating if the cache hit rate drops low sounds good, but even with the minimum `utxocachemaxsize` of 25 MiB, it stays around 75-85%.  Further, if you are setting the memory that low, it's probably because you don't have it available, so the log may not be helpful in that case.  I'm open to suggestions, though, if we want to log at a higher minimum hit ratio.

Reply:
I agree that for now it isn't super useful, but I was more thinking about the future.  Mixing is going to continue to produce more utxos, especially now that it's become a lot more accessible in the most used GUI wallet, and I expect that the block sizes will ultimately likely need to be increased.  In that climate, I could see the hit ratio naturally going down without giving the cache more memory.  That said, we developers are probably going to notice when/if bumping the default becomes necessary anyway, so I'm fine with leaving it as it is now with debug.